### PR TITLE
Improve Rust panics

### PR DESCRIPTION
**Pull request by keram88**
This improves detection and handling of panics arising in Rust programs.

### CODE REVIEW

**Comment by shaobo-he** [comment]
Quick question: panic checking is always enabled, right?

**Comment by keram88** [comment]
> Quick question: panic checking is always enabled, right?

That is the intended behavior. Do you have something in mind?

**Comment by zvonimir** [comment]
@keram88 and @shaobo-he : my hunch is that this should not always be enabled. There should be a command line option to either enable it or maybe to disable it (if it is enabled by default). No?
Otherwise is some scenarios users might get tons of false bugs. Yeah, if I think about it more, I would say that it should not be enabled by default. Thoughts?

File: lib/smack/Naming.cpp [hunk 1]
```diff
@@ -62,8 +62,11 @@ const std::string Naming::REC_MEM_OP = "boogie_si_record_mop";
 const std::string Naming::MEM_OP_VAL = "$MOP";
 
 const std::string Naming::RUST_ENTRY = "_ZN3std2rt10lang_start";
-const std::string Naming::RUST_PANIC1 = "_ZN4core9panicking5panic";
-const std::string Naming::RUST_PANIC2 = "_ZN3std9panicking11begin_panic";
+const std::vector<std::string> Naming::RUST_PANICS = {
+    "_ZN3std9panicking11begin_panic17h",
+    "_ZN3std9panicking15begin_panic_fmt17h", "_ZN4core9panicking5panic17h",
```
Comment (zvonimir) [inline lib/smack/Naming.cpp:67]: Any particular reason why these two share one line? Maybe just put them one per line.

File: lib/smack/SmackInstGenerator.cpp [hunk 2]
```diff
@@ -137,6 +137,17 @@ void SmackInstGenerator::annotate(llvm::Instruction &I, Block *B) {
   }
 }
 
+bool is_rust_panic(const std::string &name) {
```
Comment (zvonimir) [inline lib/smack/SmackInstGenerator.cpp:140]: We are using camel case for function names, no?

File: lib/smack/SmackInstGenerator.cpp [hunk 3]
```diff
@@ -640,11 +651,12 @@ void SmackInstGenerator::visitCallInst(llvm::CallInst &ci) {
...
     emit(Stmt::call(mainFunction->getName(), {}, {}));
 
-  } else if (name.find(Naming::RUST_PANIC1) != std::string::npos ||
-             name.find(Naming::RUST_PANIC2) != std::string::npos) {
+  } else if (is_rust_panic(name)) {
     // Convert Rust's panic functions into assertion violations
     emit(Stmt::assert_(Expr::lit(false),
                        {Attr::attr(Naming::RUST_PANIC_ANNOTATION)}));
+    // Block further assertions on this path
+    emit(Stmt::assume(Expr::lit(false)));
```
Comment (zvonimir) [inline lib/smack/SmackInstGenerator.cpp:659]: Why do we need this assumption? Please clarify. Thx!

File: test/rust/panic/core_panic_fail.rs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+#[start]
+fn main(_x: isize, _y: *const *const u8) -> isize {
+    panic!();
+}
+
+#[panic_handler]
+fn panic(_expr: &PanicInfo) -> !{
+    loop {}
+}
+#[lang="eh_personality"]
```
Comment (zvonimir) [inline test/rust/panic/core_panic_fail.rs:17]: Empty line missing it seems.

File: test/rust/panic/core_panic_fmt_fail.rs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+#[start]
+fn main(_x: isize, _y: *const *const u8) -> isize {
+    panic!("Something {}", 7);
+}
+
+#[panic_handler]
+fn panic(_expr: &PanicInfo) -> !{
+    loop {}
+}
+#[lang="eh_personality"]
```
Comment (zvonimir) [inline test/rust/panic/core_panic_fmt_fail.rs:17]: Empty line missing it seems.

**Comment by shaobo-he** [comment]
> @keram88 and @shaobo-he : my hunch is that this should not always be enabled. There should be a command line option to either enable it or maybe to disable it (if it is enabled by default). No?
> Otherwise is some scenarios users might get tons of false bugs. Yeah, if I think about it more, I would say that it should not be enabled by default. Thoughts?

That's my thought too. Another thing to notice is that the verifier may spend a lot of time trying to prove the panics are unreachable instead of focusing on the assertions.

File: lib/smack/Naming.cpp (hunk 1 above)
Comment (keram88) [reply lib/smack/Naming.cpp:67]: This is what clang format wanted.

**Comment by keram88** [comment]
I'll add functionality to make Rust panic opt in.

File: lib/smack/SmackInstGenerator.cpp (hunk 3 above)
Comment (keram88) [reply lib/smack/SmackInstGenerator.cpp]: I realized that this is what will happen anyway. Panics are always terminal.

**Comment by keram88** [comment]
Do we want to disable assertion checking when checking for panics?

**Pull request merged**